python/ray/tests/test_ray_shutdown.py

```diff
@@ -337,6 +341,100 @@ def get_raylet_runtime_env_agent_procs(worker):
     assert exit_code == 0
 
 
+def test_worker_sigterm(shutdown_only):
+    """Verify a worker process is killed by a sigterm."""
+    ray.init(num_cpus=1)
+
+    @ray.remote
+    def f():
+        return os.getpid(), ray.get_runtime_context().get_worker_id()
+
+    pid, worker_id = ray.get(f.remote())
+    os.kill(pid, signal.SIGTERM)
+
+    def verify():
+        w = get_worker(id=worker_id)
+        alive = w.is_alive
+        assert not alive
+        assert "SIGTERM" in w.exit_detail
+        assert w.exit_type == "INTENDED_SYSTEM_EXIT"
+        return True
+
+    wait_for_condition(verify)
+
+    @ray.remote(num_cpus=1)
+    class Actor:
+        def pid(self):
+            return os.getpid(), ray.get_runtime_context().get_worker_id()
+
+    a = Actor.remote()
+    pid, worker_id = ray.get(a.pid.remote())
+    os.kill(pid, signal.SIGTERM)
+
+    def verify():
+        w = get_worker(id=worker_id)
+        alive = w.is_alive
+        assert not alive
+        assert "SIGTERM" in w.exit_detail
+        assert w.exit_type == "INTENDED_SYSTEM_EXIT"
+        return True
+
+    wait_for_condition(verify)
+
+
+def test_worker_proc_child_no_leak(shutdown_only):
+    """Verify a worker process is not leaked when placement group is removed"""
+    ray.init(num_cpus=1)
+
+    @ray.remote(num_cpus=1)
+    class Actor:
+        def __init__(self) -> None:
+            self.p = None
+
+        def sleep(self):
+            mp_context = multiprocessing.get_context("spawn")
+            self.p = mp_context.Process(target=time.sleep, args=(1000,))
+            self.p.daemon = True
+            self.p.start()
+
+            print(f"[pid={os.getpid()}ppid={os.getppid()}]sleeping for 1")
+            time.sleep(1)
+            return ray.get_runtime_context().get_worker_id(), self.p.pid
```

Comment:
i wonder what would happen if the python actor has its own sigterm handlers? Do we need to worry about it?

Reply:
I think then it will follow the behavior user defines. I think we should not encourage users to do it (but instead encourage using a destructor). I think we need some e2e thoughts here (planning to organize here; https://docs.google.com/document/d/1mr4j5p9a0Ce5oYaUDoaiHUdOsbYfBOmAO4CPQTSjwR4/edit#heading=h.fwxjsbhjhpi5)